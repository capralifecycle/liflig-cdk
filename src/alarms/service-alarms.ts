import * as cdk from "aws-cdk-lib"
import * as cloudwatch from "aws-cdk-lib/aws-cloudwatch"
import * as logs from "aws-cdk-lib/aws-logs"
import * as constructs from "constructs"

export interface ServiceAlarmsProps extends cdk.StackProps {
  action: cloudwatch.IAlarmAction
  serviceName: string
}

/**
 * Various alarms and monitoring.
 *
 * By itself no alarms is created. Use the methods available
 * to add alarms.
 *
 * See SlackAlarm construct for SNS Action.
 */
export class ServiceAlarms extends constructs.Construct {
  private readonly action: cloudwatch.IAlarmAction
  private readonly serviceName: string

  constructor(
    scope: constructs.Construct,
    id: string,
    props: ServiceAlarmsProps,
  ) {
    super(scope, id)

    this.action = props.action
    this.serviceName = props.serviceName
  }

  /**
   * For logs stored as JSON, monitor log entries logged
   * with level ERROR or higher, as well as any requests
   * that causes 500 for logging with liflig-logging.
   */
  addJsonErrorAlarm(props: {
    logGroup: logs.ILogGroup
    alarmDescription?: string
    /** Set to `false` to stop the alarm from sending OK events.
     * @default true */
    enableOkAction?: boolean
  }): void {
    const errorMetricFilter = props.logGroup.addMetricFilter(
      "ErrorMetricFilter",
      {
        filterPattern: logs.FilterPattern.any(
          logs.FilterPattern.stringValue("$.level", "=", "ERROR"),
          // FATAL covers some applications we run that uses log4j or
          // other libraries. It is not existent in slf4j.
          logs.FilterPattern.stringValue("$.level", "=", "FATAL"),
          logs.FilterPattern.stringValue(
            // For liflig-logging.
            "$.requestInfo.status.code",
            "=",
            "INTERNAL_SERVER_ERROR",
          ),
        ),
        metricName: "Errors",
        metricNamespace: `stack/${cdk.Stack.of(this).stackName}/${
          this.serviceName
        }/Errors`,
      },
    )

    const errorAlarm = errorMetricFilter
      .metric()
      .with({
        statistic: "Sum",
        period: cdk.Duration.seconds(60),
      })
      .createAlarm(this, "ErrorLogAlarm", {
        alarmDescription:
          props.alarmDescription ?? `${this.serviceName} logged an error`,
        evaluationPeriods: 1,
        threshold: 1,
        treatMissingData: cloudwatch.TreatMissingData.NOT_BREACHING,
      })

    errorAlarm.addAlarmAction(this.action)
    if (props.enableOkAction ?? true) {
      errorAlarm.addOkAction(this.action)
    }
  }

  /**
   * Monitor healthy host count, 5xx status codes and connection errors from load balancer.
   */
  addTargetGroupAlarm(props: {
    targetGroupFullName: string
    loadBalancerFullName: string
    tooMany5xxResponsesFromTargetsAlarmOverride?: {
      period: cdk.Duration
      evaluationPeriods: number
      threshold: number
    }
  }): void {
    const healthAlarm = new cloudwatch.Metric({
      metricName: "HealthyHostCount",
      namespace: "AWS/ApplicationELB",
      statistic: "Average",
      period: cdk.Duration.seconds(60),
      dimensionsMap: {
        TargetGroup: props.targetGroupFullName,
        LoadBalancer: props.loadBalancerFullName,
      },
    }).createAlarm(this, "HealthAlarm", {
      alarmDescription:
        "Service might be unavailable! It is not responding to health checks.",
      comparisonOperator: cloudwatch.ComparisonOperator.LESS_THAN_THRESHOLD,
      evaluationPeriods: 1,
      threshold: 1,
      treatMissingData: cloudwatch.TreatMissingData.BREACHING,
    })

    healthAlarm.addAlarmAction(this.action)
    healthAlarm.addOkAction(this.action)

    const connectionAlarm = new cloudwatch.Metric({
      metricName: "TargetConnectionErrorCount",
      namespace: "AWS/ApplicationELB",
      statistic: "Sum",
      period: cdk.Duration.seconds(60),
      dimensionsMap: {
        TargetGroup: props.targetGroupFullName,
        LoadBalancer: props.loadBalancerFullName,
      },
    }).createAlarm(this, "ConnectionAlarm", {
      actionsEnabled: true,
      alarmDescription: "Load balancer could not connect to target",
      evaluationPeriods: 1,
      threshold: 1,
      treatMissingData: cloudwatch.TreatMissingData.NOT_BREACHING,
    })

    connectionAlarm.addAlarmAction(this.action)
    connectionAlarm.addOkAction(this.action)

    const tooMany5xxResponsesFromTargetsAlarm = new cloudwatch.Metric({
      metricName: "HTTPCode_Target_5XX_Count",
      namespace: "AWS/ApplicationELB",
      statistic: "Sum",
      period:
        props.tooMany5xxResponsesFromTargetsAlarmOverride?.period ??
        cdk.Duration.seconds(60),
      dimensionsMap: {
        TargetGroup: props.targetGroupFullName,
        LoadBalancer: props.loadBalancerFullName,
      },
    }).createAlarm(this, "AlbTargets5xxAlarm", {
      actionsEnabled: true,
      alarmDescription:
        "Load balancer received too many 5XX response codes from its targets.",
      evaluationPeriods:
        props.tooMany5xxResponsesFromTargetsAlarmOverride?.evaluationPeriods ??
        1,
      threshold:
        props.tooMany5xxResponsesFromTargetsAlarmOverride?.threshold ?? 10,
      treatMissingData: cloudwatch.TreatMissingData.NOT_BREACHING,
    })

    tooMany5xxResponsesFromTargetsAlarm.addAlarmAction(this.action)
    tooMany5xxResponsesFromTargetsAlarm.addOkAction(this.action)
  }

  /**
   * Monitor 5xx status codes produced by load balancer.
   */
  addAlbAlarms(props: {
    targetGroupFullName: string
    loadBalancerFullName: string
    tooMany5xxResponsesFromAlbAlarmOverride?: {
      period: cdk.Duration
      evaluationPeriods: number
      threshold: number
    }
  }): void {
    const tooMany5xxResponsesFromAlbAlarm = new cloudwatch.Metric({
      metricName: "HTTPCode_ELB_5XX_Count",
      namespace: "AWS/ApplicationELB",
      statistic: "Sum",
      period:
        props.tooMany5xxResponsesFromAlbAlarmOverride?.period ??
        cdk.Duration.seconds(60),
      dimensionsMap: {
        TargetGroup: props.targetGroupFullName,
        LoadBalancer: props.loadBalancerFullName,
      },
    }).createAlarm(this, "Alb5xxAlarm", {
      actionsEnabled: true,
      alarmDescription: "Load balancer returns too many 5XX response codes.",
      evaluationPeriods:
        props.tooMany5xxResponsesFromAlbAlarmOverride?.evaluationPeriods ?? 1,
      threshold: props.tooMany5xxResponsesFromAlbAlarmOverride?.threshold ?? 1,
      treatMissingData: cloudwatch.TreatMissingData.NOT_BREACHING,
    })

    tooMany5xxResponsesFromAlbAlarm.addAlarmAction(this.action)
    tooMany5xxResponsesFromAlbAlarm.addOkAction(this.action)
  }
}
