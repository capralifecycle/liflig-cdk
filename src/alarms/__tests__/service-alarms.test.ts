import "@aws-cdk/assert/jest"
import * as cloudwatchActions from "@aws-cdk/aws-cloudwatch-actions"
import * as logs from "@aws-cdk/aws-logs"
import * as sns from "@aws-cdk/aws-sns"
import { App, Stack } from "@aws-cdk/core"
import "jest-cdk-snapshot"
import { ServiceAlarms } from "../service-alarms"

test("create alarms", () => {
  const app = new App()
  const supportStack = new Stack(app, "SupportStack")
  const stack = new Stack(app, "Stack")

  const topic = new sns.Topic(supportStack, "Topic")
  const snsAction = new cloudwatchActions.SnsAction(topic)

  const logGroup = new logs.LogGroup(stack, "LogGroup")

  const alarms = new ServiceAlarms(stack, "ServiceAlarms", {
    serviceName: "service-name",
    snsAction,
  })

  alarms.addJsonErrorAlarm({
    logGroup,
  })

  alarms.addTargetGroupAlarm({
    loadBalancerFullName: "app/my-load-balancer/50dc6c495c0c9188",
    targetGroupFullName: "targetgroup/my-target-group/cbf133c568e0d028",
  })

  expect(stack).toMatchCdkSnapshot()
})
