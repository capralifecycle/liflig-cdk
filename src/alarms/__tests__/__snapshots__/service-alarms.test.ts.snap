// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create alarms 1`] = `
Object {
  "Resources": Object {
    "LogGroupErrorMetricFilter551B92D2": Object {
      "Properties": Object {
        "FilterPattern": "{ ($.level = \\"ERROR\\") || ($.level = \\"FATAL\\") || ($.requestInfo.status.code = \\"INTERNAL_SERVER_ERROR\\") }",
        "LogGroupName": Object {
          "Ref": "LogGroupF5B46931",
        },
        "MetricTransformations": Array [
          Object {
            "MetricName": "Errors",
            "MetricNamespace": "stack/Stack/service-name/Errors",
            "MetricValue": "1",
          },
        ],
      },
      "Type": "AWS::Logs::MetricFilter",
    },
    "LogGroupF5B46931": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "ServiceAlarmsAlbTargets5xxAlarm55A1710E": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmActions": Array [
          Object {
            "Fn::ImportValue": "SupportStack:ExportsOutputRefTopicBFC7AF6ECB4A357A",
          },
        ],
        "AlarmDescription": "Load balancer received too many 5XX responses from target(s) in ECS service 'service-name'.",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "LoadBalancer",
            "Value": "app/my-load-balancer/50dc6c495c0c9188",
          },
          Object {
            "Name": "TargetGroup",
            "Value": "targetgroup/my-target-group/cbf133c568e0d028",
          },
        ],
        "EvaluationPeriods": 3,
        "MetricName": "HTTPCode_Target_5XX_Count",
        "Namespace": "AWS/ApplicationELB",
        "OKActions": Array [
          Object {
            "Fn::ImportValue": "SupportStack:ExportsOutputRefTopicBFC7AF6ECB4A357A",
          },
        ],
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 10,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ServiceAlarmsConnectionAlarm1F083C01": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Load balancer is failing to connect to target(s) in ECS service 'service-name'.",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "LoadBalancer",
            "Value": "app/my-load-balancer/50dc6c495c0c9188",
          },
          Object {
            "Name": "TargetGroup",
            "Value": "targetgroup/my-target-group/cbf133c568e0d028",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "TargetConnectionErrorCount",
        "Namespace": "AWS/ApplicationELB",
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ServiceAlarmsErrorLogAlarmEBEFC89B": Object {
      "Properties": Object {
        "AlarmActions": Array [
          Object {
            "Fn::ImportValue": "SupportStack:ExportsOutputRefTopicBFC7AF6ECB4A357A",
          },
        ],
        "AlarmDescription": "service-name logged an error",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "EvaluationPeriods": 1,
        "MetricName": "Errors",
        "Namespace": "stack/Stack/service-name/Errors",
        "OKActions": Array [
          Object {
            "Fn::ImportValue": "SupportStack:ExportsOutputRefTopicBFC7AF6ECB4A357A",
          },
        ],
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ServiceAlarmsHealthAlarm888BAAB5": Object {
      "Properties": Object {
        "AlarmDescription": "There are no healthy target(s) in ECS service 'service-name'.",
        "ComparisonOperator": "LessThanThreshold",
        "Dimensions": Array [
          Object {
            "Name": "LoadBalancer",
            "Value": "app/my-load-balancer/50dc6c495c0c9188",
          },
          Object {
            "Name": "TargetGroup",
            "Value": "targetgroup/my-target-group/cbf133c568e0d028",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "HealthyHostCount",
        "Namespace": "AWS/ApplicationELB",
        "Period": 60,
        "Statistic": "Minimum",
        "Threshold": 1,
        "TreatMissingData": "breaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ServiceAlarmsTargetHealthAlarm18599060": Object {
      "Properties": Object {
        "ActionsEnabled": false,
        "AlarmActions": Array [
          Object {
            "Fn::ImportValue": "SupportStack:ExportsOutputRefTopicBFC7AF6ECB4A357A",
          },
        ],
        "AlarmDescription": "The load balancer is either receiving bad health checks from or is unable to connect to target(s) in ECS service 'service-name'",
        "AlarmName": "StackServiceAlarmsTargetHealthAlarm30873D46",
        "AlarmRule": Object {
          "Fn::Join": Array [
            "",
            Array [
              "(ALARM(\\"",
              Object {
                "Fn::GetAtt": Array [
                  "ServiceAlarmsConnectionAlarm1F083C01",
                  "Arn",
                ],
              },
              "\\") OR ALARM(\\"",
              Object {
                "Fn::GetAtt": Array [
                  "ServiceAlarmsHealthAlarm888BAAB5",
                  "Arn",
                ],
              },
              "\\"))",
            ],
          ],
        },
        "OKActions": Array [
          Object {
            "Fn::ImportValue": "SupportStack:ExportsOutputRefTopicBFC7AF6ECB4A357A",
          },
        ],
      },
      "Type": "AWS::CloudWatch::CompositeAlarm",
    },
    "ServiceAlarmsTargetResponseTimeAlarmA0DC0898": Object {
      "Properties": Object {
        "AlarmActions": Array [
          Object {
            "Fn::ImportValue": "SupportStack:ExportsOutputRefTopicBFC7AF6ECB4A357A",
          },
        ],
        "AlarmDescription": "5% of responses from ECS service 'service-name' are taking longer than expected.",
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": Array [
          Object {
            "Name": "LoadBalancer",
            "Value": "app/my-load-balancer/50dc6c495c0c9188",
          },
          Object {
            "Name": "TargetGroup",
            "Value": "targetgroup/my-target-group/cbf133c568e0d028",
          },
        ],
        "EvaluationPeriods": 1,
        "ExtendedStatistic": "p95",
        "MetricName": "TargetResponseTime",
        "Namespace": "AWS/ApplicationELB",
        "OKActions": Array [
          Object {
            "Fn::ImportValue": "SupportStack:ExportsOutputRefTopicBFC7AF6ECB4A357A",
          },
        ],
        "Period": 300,
        "Threshold": 0.5,
        "TreatMissingData": "ignore",
        "Unit": "Seconds",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
}
`;
